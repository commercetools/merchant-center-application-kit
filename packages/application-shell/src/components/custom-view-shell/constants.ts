// TODO: This file is duplicated (also exists in application-components custom views components)
export const CUSTOM_VIEWS_EVENTS_NAMES = {
  CUSTOM_VIEW_BOOTSTRAP: 'custom-view-bootstrap',
  CUSTOM_VIEW_INITIALIZATION: 'custom-view-initialization',
};

export const CUSTOM_VIEWS_EVENTS_META = {
  SOURCE: 'mc-host-application',
  DESTINATION_PREFIX: 'custom-view-',
};
