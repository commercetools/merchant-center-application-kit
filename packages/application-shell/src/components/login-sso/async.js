import React from 'react';

export default React.lazy(() =>
  import('./index' /* webpackChunkName: "login-sso" */)
);
