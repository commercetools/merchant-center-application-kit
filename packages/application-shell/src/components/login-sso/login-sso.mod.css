.error {
  color: var(--color-red);
}
