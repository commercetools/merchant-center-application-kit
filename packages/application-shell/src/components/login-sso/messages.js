import { defineMessages } from 'react-intl';

export default defineMessages({
  title: {
    id: 'Login.sso.title',
    defaultMessage: 'Sign in to your Organization',
  },
  organization: {
    id: 'Login.sso.organization',
    defaultMessage: "Enter your Organization's name to sign in",
  },
  caseSensitive: {
    id: 'Login.sso.caseSensitive',
    defaultMessage: 'This field is case sensitive',
  },
  proceed: {
    id: 'Login.sso.proceed',
    defaultMessage: 'Proceed to sign in',
  },
  organizationGeneralError: {
    id: 'Login.sso.organizationGeneralError',
    defaultMessage: 'Organization name is not found',
  },
  organizationGeneralErrorSecondary: {
    id: 'Login.sso.organizationGeneralErrorSecondary',
    defaultMessage:
      'Either your Organization is not using SSO, is misspelled or does not exist. Try re-entering your Organization and contact your system administrator if you have any further questions.',
  },
});
