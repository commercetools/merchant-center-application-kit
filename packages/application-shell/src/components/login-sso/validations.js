import omitEmpty from 'omit-empty-es';
import { TextInput } from '@commercetools-frontend/ui-kit';

// eslint-disable-next-line import/prefer-default-export
export const validate = values => {
  const errorsByField = {
    organizationName: {},
  };

  if (TextInput.isEmpty(values.organizationName)) {
    errorsByField.organizationName.missing = true;
  }

  return omitEmpty(errorsByField);
};
