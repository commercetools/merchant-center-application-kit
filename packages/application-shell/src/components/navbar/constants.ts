export const NAVBAR = {
  expanderHeight: '50px',
  expanderSize: '40px',
  iconSize: '24px',
  iconSizeHover: '28px',
  itemSize: '48px',
  itemHeight: '56px',
  sublistIndentationWhenCollapsed: '72px',
  sublistIndentationWhenExpanded: '248px',
  sublistWidth: '253px',
  leftNavigationTransition: 'all 150ms cubic-bezier(1, 0, 0.58, 1)',
  widthLeftNavigation: '80px',
  widthLeftNavigationWhenExpanded: '256px',
} as const;
