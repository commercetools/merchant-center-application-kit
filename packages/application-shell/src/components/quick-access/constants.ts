// eslint-disable-next-line import/prefer-default-export

export const permissions = {
  ViewOrders: 'ViewOrders',
  ManageOrders: 'ManageOrders',
  ViewProducts: 'ViewProducts',
  ManageProducts: 'ManageProducts',
  ViewCategories: 'ViewCategories',
  ManageCategories: 'ManageCategories',
  ViewCustomers: 'ViewCustomers',
  ManageCustomers: 'ManageCustomers',
  ViewCustomerGroups: 'ViewCustomerGroups',
  ManageCustomerGroups: 'ManageCustomerGroups',
  ViewProductDiscounts: 'ViewProductDiscounts',
  ManageProductDiscounts: 'ManageProductDiscounts',
  ViewDiscountCodes: 'ViewDiscountCodes',
  ManageDiscountCodes: 'ManageDiscountCodes',
  ViewCartDiscounts: 'ViewCartDiscounts',
  ManageCartDiscounts: 'ManageCartDiscounts',
  ViewProjectSettings: 'ViewProjectSettings',
  ManageProjectSettings: 'ManageProjectSettings',
  ViewDeveloperSettings: 'ViewDeveloperSettings',
  ManageDeveloperSettings: 'ManageDeveloperSettings',
  ViewProductTypes: 'ViewProductTypes',
  ManageProductTypes: 'ManageProductTypes',
} as const;
