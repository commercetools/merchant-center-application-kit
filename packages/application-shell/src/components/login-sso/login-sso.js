import querystring from 'querystring';
import React from 'react';
import PropTypes from 'prop-types';
import { FormattedMessage, injectIntl } from 'react-intl';
import { compose } from 'recompose';
import { Formik } from 'formik';
import uuid from 'uuid/v4';
import {
  Text,
  PrimaryButton,
  Spacings,
  TextField,
  ContentNotification,
} from '@commercetools-frontend/ui-kit';
import {
  joinPaths,
  trimLeadingAndTrailingSlashes,
} from '@commercetools-frontend/url-utils';
import { actions as sdkActions } from '@commercetools-frontend/sdk';
import { connect } from 'react-redux';
import { ORGANIZATION_GENERAL_ERROR, STORAGE_KEYS } from '../../constants';
import PublicPageContainer from '../public-page-container';
import LoginBox from '../login-box';
import messages from './messages';
import { validate } from './validations';

export const getMessageKeyForError = error => {
  switch (error) {
    case ORGANIZATION_GENERAL_ERROR:
      return 'organizationGeneralError';
    default:
      return undefined;
  }
};

function generateAndCacheNonceWithState(state) {
  const nonce = uuid();
  // We store additional information within the given `nonce`
  // to then retrieve it later when the IdP redirects back
  // to our application. The URL will contain the `nonce` within
  // the id_token and, once validated, we can retrieve and use
  // the state object.
  // https://auth0.com/docs/protocols/oauth2/oauth-state#how-to-use-the-parameter-to-restore-application-state
  window.sessionStorage.setItem(
    `${STORAGE_KEYS.NONCE}_${nonce}`,
    JSON.stringify(state)
  );
  return nonce;
}

const originUrl = window.location.origin;

const redirect = targetUrl => window.location.replace(targetUrl);

export class LoginSSO extends React.PureComponent {
  static displayName = 'LoginSSO';
  static propTypes = {
    match: PropTypes.shape({
      url: PropTypes.string.isRequired,
    }).isRequired,
    intl: PropTypes.shape({
      formatMessage: PropTypes.func.isRequired,
    }).isRequired,
    getOrganizationByName: PropTypes.func.isRequired,
  };
  handleSubmit = (values, actions) => {
    this.props.getOrganizationByName(values.organizationName).then(
      authProvider => {
        actions.setSubmitting(false);

        // Quick note: we assume that the authorization endpoint is /authorize
        // This endpoint name is not mandatory. However, it is used as a common
        // practice throughout the OIDC specification.
        // We might have to let customers set a custom endpoint in the future.
        const params = querystring.stringify({
          scope: 'openid email profile',
          response_type: 'id_token',
          client_id: authProvider.clientId,
          redirect_uri: trimLeadingAndTrailingSlashes(
            // Avoid providing query parameters as some IdP (e.g. Azure) apparently
            // will consider the full URL to match as part of the callback whitelist.
            // Instead, we store additional information within the `nonce` value
            // which is stored in sessionStorage. See `generateAndCacheNonceWithState`.
            joinPaths(originUrl, this.props.match.url, `callback`)
          ),
          nonce: generateAndCacheNonceWithState({
            organizationId: authProvider.organizationId,
          }),
        });
        redirect(`${authProvider.authorizeUrl}?${params}`);
      },
      error => {
        actions.setSubmitting(false);
        actions.setStatus({
          errorMessage:
            error.statusCode === 404
              ? ORGANIZATION_GENERAL_ERROR
              : error.message,
        });
      }
    );
  };
  render() {
    return (
      <PublicPageContainer>
        <LoginBox>
          <Formik
            initialValues={{ organizationName: '' }}
            validate={validate}
            onSubmit={this.handleSubmit}
            render={formikProps => {
              const errorMessageKey = getMessageKeyForError(
                formikProps.status && formikProps.status.errorMessage
              );
              return (
                <form onSubmit={formikProps.handleSubmit}>
                  <Spacings.Stack scale="m">
                    {formikProps.status && formikProps.status.errorMessage ? (
                      <ContentNotification type="error">
                        <Spacings.Stack scale="s">
                          <Text.Body>
                            {messages[errorMessageKey] ? (
                              <FormattedMessage
                                {...messages[errorMessageKey]}
                              />
                            ) : (
                              formikProps.status.errorMessage
                            )}
                          </Text.Body>
                          {messages[`${errorMessageKey}Secondary`] ? (
                            <Text.Body>
                              <FormattedMessage
                                {...messages[`${errorMessageKey}Secondary`]}
                              />
                            </Text.Body>
                          ) : null}
                        </Spacings.Stack>
                      </ContentNotification>
                    ) : null}
                    <Text.Subheadline elementType="h4">
                      <FormattedMessage {...messages.title} />
                    </Text.Subheadline>
                    <TextField
                      name="organizationName"
                      title={this.props.intl.formatMessage(
                        messages.organization
                      )}
                      hint={this.props.intl.formatMessage(
                        messages.caseSensitive
                      )}
                      isAutofocussed={true}
                      isRequired={true}
                      value={formikProps.values.organizationName}
                      touched={formikProps.touched.organizationName}
                      errors={formikProps.errors.organizationName}
                      onChange={formikProps.handleChange}
                      onBlur={formikProps.handleBlur}
                      isDisabled={formikProps.isSubmitting}
                    />
                    <Spacings.Inline>
                      <PrimaryButton
                        type="submit"
                        label={this.props.intl.formatMessage(messages.proceed)}
                        onClick={formikProps.handleSubmit}
                        isDisabled={formikProps.isSubmitting}
                      />
                    </Spacings.Inline>
                  </Spacings.Stack>
                </form>
              );
            }}
          />
        </LoginBox>
      </PublicPageContainer>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  getOrganizationByName: name =>
    dispatch(sdkActions.get({ uri: `/sso/organization?name=${name}` })),
});

export default compose(
  injectIntl,
  connect(
    null,
    mapDispatchToProps
  )
)(LoginSSO);
