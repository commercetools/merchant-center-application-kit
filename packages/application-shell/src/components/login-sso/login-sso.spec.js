import React from 'react';
import { shallow } from 'enzyme';
import { Formik } from 'formik';
import { PrimaryButton, Text } from '@commercetools-frontend/ui-kit';
import { ORGANIZATION_GENERAL_ERROR } from '../../constants';
import { LoginSSO, getMessageKeyForError } from './login-sso';

const createTestProps = props => ({
  match: {
    url: '/login/sso',
  },
  intl: {
    formatMessage: jest.fn(message => message.id),
  },
  originUrl: 'http://mc.ct.com',
  redirectTo: jest.fn(),
  getOrganizationByName: jest.fn(),
  ...props,
});

const createTestPropsForFormik = props => ({
  values: { organizationName: '' },
  errors: {},
  status: undefined,
  touched: {},
  isValid: false,
  handleSubmit: jest.fn(),
  isSubmitting: false,
  setFieldValue: jest.fn(),
  setFieldTouched: jest.fn(),
  ...props,
});

jest.mock('uuid/v4', () => () => 'foo-uuid');
jest.mock('@commercetools-frontend/storage');

describe('rendering', () => {
  let wrapper;
  let props;
  beforeEach(() => {
    props = createTestProps();
    wrapper = shallow(<LoginSSO {...props} />);
  });

  it('should render a LoginBox', () => {
    expect(wrapper).toRender('LoginBox');
  });
  it('should render <Formik>', () => {
    expect(wrapper).toRender(Formik);
  });
  describe('<Formik>', () => {
    it('should pass initial form values', () => {
      expect(wrapper.find(Formik)).toHaveProp(
        'initialValues',
        wrapper.instance().initialFormValues
      );
    });
    it('should disable validation on blur', () => {
      expect(wrapper.find(Formik)).toHaveProp('validateOnBlur', false);
    });
    describe('form elements', () => {
      let formikProps;
      let formikRenderWrapper;
      beforeEach(() => {
        formikProps = createTestPropsForFormik();
        formikRenderWrapper = shallow(
          wrapper.find(Formik).prop('render')(formikProps)
        );
      });
      it('should match layout', () => {
        expect(formikRenderWrapper).toMatchSnapshot();
      });
      it('should pass organizationName value to Field', () => {
        expect(
          formikRenderWrapper.find({ name: 'organizationName' })
        ).toHaveProp('value', '');
      });
      describe('when organizationName is not valid', () => {
        beforeEach(() => {
          formikProps = createTestPropsForFormik({
            touched: { organizationName: true },
            errors: { organizationName: 'Required field' },
          });
          formikRenderWrapper = shallow(
            wrapper.find(Formik).prop('render')(formikProps)
          );
        });
        it('should render error message', () => {
          expect(formikRenderWrapper).toContainReact(
            <p className="error">{'Required field'}</p>
          );
        });
      });
      describe('general error message', () => {
        describe('when form status contains error message', () => {
          describe('when error message is matching the key ORGANIZATION_GENERAL_ERROR', () => {
            beforeEach(() => {
              formikProps = createTestPropsForFormik({
                status: { errorMessage: ORGANIZATION_GENERAL_ERROR },
              });
              formikRenderWrapper = shallow(
                wrapper.find(Formik).prop('render')(formikProps)
              );
            });
            it('should render <Notification>', () => {
              expect(formikRenderWrapper).toRender('Notification');
            });
            it('should render message with matching key', () => {
              expect(formikRenderWrapper).toRender({
                id: 'Login.sso.organizationGeneralError',
              });
            });
            it('should render secondary message with matching key', () => {
              expect(formikRenderWrapper).toRender({
                id: 'Login.sso.organizationGeneralErrorSecondary',
              });
            });
          });
          describe('when error message is not matching the key ORGANIZATION_GENERAL_ERROR', () => {
            beforeEach(() => {
              formikProps = createTestPropsForFormik({
                status: { errorMessage: 'Some other error' },
              });
              formikRenderWrapper = shallow(
                wrapper.find(Formik).prop('render')(formikProps)
              );
            });
            it('should render <Notification>', () => {
              expect(formikRenderWrapper).toRender('Notification');
            });
            it('should render error message as-is', () => {
              expect(formikRenderWrapper).toContainReact(
                <Text.Body>{'Some other error'}</Text.Body>
              );
            });
            it('should not render secondary message', () => {
              expect(formikRenderWrapper).not.toRender({
                id: 'Login.sso.organizationGeneralErrorSecondary',
              });
            });
          });
        });
      });
      describe('submit button', () => {
        describe('when form is not valid', () => {
          beforeEach(() => {
            formikProps = createTestPropsForFormik({
              isValid: false,
            });
            formikRenderWrapper = shallow(
              wrapper.find(Formik).prop('render')(formikProps)
            );
          });
          it('should disable button', () => {
            expect(formikRenderWrapper.find(PrimaryButton)).toHaveProp(
              'isDisabled',
              true
            );
          });
        });
        describe('when form is valid', () => {
          describe('when form is submitting', () => {
            beforeEach(() => {
              formikProps = createTestPropsForFormik({
                isValid: true,
                isSubmitting: true,
              });
              formikRenderWrapper = shallow(
                wrapper.find(Formik).prop('render')(formikProps)
              );
            });
            it('should disable button', () => {
              expect(formikRenderWrapper.find(PrimaryButton)).toHaveProp(
                'isDisabled',
                true
              );
            });
          });
          describe('when form is not submitting', () => {
            beforeEach(() => {
              formikProps = createTestPropsForFormik({
                isValid: true,
                isSubmitting: false,
              });
              formikRenderWrapper = shallow(
                wrapper.find(Formik).prop('render')(formikProps)
              );
            });
            it('should enable button', () => {
              expect(formikRenderWrapper.find(PrimaryButton)).toHaveProp(
                'isDisabled',
                false
              );
            });
          });
        });
      });
    });
  });
});

describe('interaction', () => {
  let wrapper;
  let props;
  describe('submitting the form', () => {
    const formActions = { setSubmitting: jest.fn(), setStatus: jest.fn() };
    describe('when request is successful', () => {
      describe('when authProvider protocol is "oidc"', () => {
        beforeEach(() => {
          window.sessionStorage = jest.fn();
          props = createTestProps({
            getOrganizationByName: jest.fn(() =>
              Promise.resolve({
                authorizeUrl: 'https://auth0.ct.com/authorize',
                organizationId: 'o1',
                clientId: '123',
              })
            ),
          });
          wrapper = shallow(<LoginSSO {...props} />);
          wrapper
            .instance()
            .handleSubmit({ organizationName: 'commercetools' }, formActions);
        });
        it('should reset submitting state', () => {
          expect(formActions.setSubmitting).toHaveBeenCalledWith(false);
        });
        it('should pass organization name to getOrganizationByName', () => {
          expect(props.getOrganizationByName).toHaveBeenCalledWith(
            'commercetools'
          );
        });
        it('should build authorize URL', () => {
          expect(props.redirectTo).toHaveBeenCalledWith(
            expect.stringContaining('https://auth0.ct.com/authorize')
          );
        });
        it('should build authorize URL with scope param', () => {
          expect(props.redirectTo).toHaveBeenCalledWith(
            expect.stringContaining(
              `scope=${encodeURIComponent('openid email profile')}`
            )
          );
        });
        it('should build authorize URL with response_type param', () => {
          expect(props.redirectTo).toHaveBeenCalledWith(
            expect.stringContaining('response_type=id_token')
          );
        });
        it('should build authorize URL with client_id param', () => {
          expect(props.redirectTo).toHaveBeenCalledWith(
            expect.stringContaining('client_id=123')
          );
        });
        it('should save state in nonce', () => {
          expect(window.sessionStorage.setItem).toHaveBeenCalledWith(
            expect.stringContaining('nonce_foo-uuid'),
            JSON.stringify({ organizationId: 'o1' })
          );
        });
        it('should build authorize URL with nonce param', () => {
          expect(props.redirectTo).toHaveBeenCalledWith(
            expect.stringContaining('nonce=foo-uuid')
          );
        });
      });
    });
    describe('when request fails', () => {
      describe('when status code is 404', () => {
        beforeEach(() => {
          props = createTestProps({
            getOrganizationByName: jest.fn(() => {
              const error = new Error('Oops');
              error.statusCode = 404;
              return Promise.reject(error);
            }),
          });
          wrapper = shallow(<LoginSSO {...props} />);
          wrapper
            .instance()
            .handleSubmit({ organizationName: 'commercetools' }, formActions);
        });
        it('should reset submitting state', () => {
          expect(formActions.setSubmitting).toHaveBeenCalledWith(false);
        });
        it('should set form status with error message', () => {
          expect(formActions.setStatus).toHaveBeenCalledWith({
            errorMessage: ORGANIZATION_GENERAL_ERROR,
          });
        });
      });
      describe('when status code is not 404', () => {
        beforeEach(() => {
          props = createTestProps({
            getOrganizationByName: jest.fn(() => {
              const error = new Error('Oops');
              error.statusCode = 400;
              return Promise.reject(error);
            }),
          });
          wrapper = shallow(<LoginSSO {...props} />);
          wrapper
            .instance()
            .handleSubmit({ organizationName: 'commercetools' }, formActions);
        });
        it('should reset submitting state', () => {
          expect(formActions.setSubmitting).toHaveBeenCalledWith(false);
        });
        it('should set form status with error message', () => {
          expect(formActions.setStatus).toHaveBeenCalledWith({
            errorMessage: 'Oops',
          });
        });
      });
    });
  });
});

describe('validation', () => {
  let wrapper;
  let props;
  beforeEach(() => {
    props = createTestProps();
    wrapper = shallow(<LoginSSO {...props} />);
  });
  describe('missing organization name', () => {
    it('should return an error', () => {
      expect(wrapper.instance().handleValidation({})).toEqual({
        organizationName: 'Validation.required',
      });
    });
  });
});

describe('getMessageKeyForError', () => {
  describe('ORGANIZATION_GENERAL_ERROR error', () => {
    it('should map to organizationGeneralError', () => {
      expect(getMessageKeyForError(ORGANIZATION_GENERAL_ERROR)).toBe(
        'organizationGeneralError'
      );
    });
  });
});
