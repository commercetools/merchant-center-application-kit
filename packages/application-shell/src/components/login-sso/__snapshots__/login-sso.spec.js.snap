// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering <Formik> form elements should match layout 1`] = `
<form
  onSubmit={[MockFunction]}
>
  <Stack
    alignItems="stretch"
    scale="m"
  >
    <TextSubheadline
      elementType="h3"
    >
      <FormattedMessage
        defaultMessage="Sign in to your Organization"
        id="Login.sso.title"
        values={Object {}}
      />
    </TextSubheadline>
    <TextField
      hint="Login.sso.caseSensitive"
      horizontalConstraint="scale"
      isAutofocussed={true}
      isDisabled={false}
      isRequired={true}
      name="organizationName"
      title="Login.sso.organization"
      value=""
    />
    <Inline
      alignItems="flexStart"
      scale="s"
    >
      <PrimaryButton
        isDisabled={false}
        isToggleButton={false}
        label="Login.sso.proceed"
        onClick={[MockFunction]}
        size="big"
        tone="primary"
      />
    </Inline>
  </Stack>
</form>
`;
